### PR TITLE
cmd/gitserver: add p4-fusion installation to gitserver Docker image

### PR DESCRIPTION
**Pull request by sashaostrikov**
I ended up installing p4-fusion directly in resulting image, because using latest alpine and git and building `p4-fusion` as a stage and copying the files to resulting container seem to lose some files which are needed for successful `p4-fusion` launch.

Container size increased from 570 MB to 688 MB

Closes https://github.com/sourcegraph/sourcegraph/issues/26926

### CODE REVIEW

File: cmd/gitserver/Dockerfile [hunk 1]
```diff
@@ -55,6 +56,36 @@ COPY --from=coursier /usr/local/bin/coursier /usr/local/bin/coursier
 ADD https://github.com/jtilander/p4d/raw/4600d741720f85d77852dcca7c182e96ad613358/lib/lib-x64.tgz /
 RUN tar zxf /lib-x64.tgz --directory /
 
+# p4-fusion installation
+# Fetching p4 sources archive
+RUN wget https://github.com/salesforce/p4-fusion/archive/refs/tags/v1.2.zip && \
```
Comment (ryanslade) [inline cmd/gitserver/Dockerfile:61]: We want to use the 1.3 release now

File: cmd/gitserver/Dockerfile [hunk 2]
```diff
@@ -55,6 +56,36 @@ COPY --from=coursier /usr/local/bin/coursier /usr/local/bin/coursier
...
+WORKDIR /usr/local/bin/p4-fusion-1.2
+
+# Build p4-fusion
+RUN ./generate_cache.sh Release
+RUN ./build.sh
+
+WORKDIR /usr/local/bin
+# Move exe file to /usr/local/bin where other executives are located
+RUN mv ./p4-fusion-1.2/build/p4-fusion/p4-fusion ./ && rm -rf /usr/local/bin/p4-fusion-1.2
+# p4-fusion installation completed here
```
Comment (ryanslade) [inline cmd/gitserver/Dockerfile:87]: Maybe we can delete everything in `/usr/loca/bin/p4-fusion-1.2` to free up space? We only need to keep the binary

File: cmd/gitserver/Dockerfile [hunk 3]
```diff
@@ -55,6 +56,36 @@ COPY --from=coursier /usr/local/bin/coursier /usr/local/bin/coursier
...
+    tar -C /usr/local/bin/p4-fusion-1.2/vendor/helix-core-api/linux -xzvf /usr/local/bin/p4-fusion-1.2/vendor/helix-core-api/linux/p4api.tgz --strip 1
+
+WORKDIR /usr/local/bin/p4-fusion-1.2
+
+# Build p4-fusion
+RUN ./generate_cache.sh Release
+RUN ./build.sh
+
+WORKDIR /usr/local/bin
+# Move exe file to /usr/local/bin where other executives are located
```
Comment (ryanslade) [inline cmd/gitserver/Dockerfile:85]:
```suggestion
# Move exe file to /usr/local/bin where other executables are located
```

File: cmd/gitserver/Dockerfile [hunk 4]
```diff
@@ -55,6 +56,36 @@ COPY --from=coursier /usr/local/bin/coursier /usr/local/bin/coursier
 ADD https://github.com/jtilander/p4d/raw/4600d741720f85d77852dcca7c182e96ad613358/lib/lib-x64.tgz /
 RUN tar zxf /lib-x64.tgz --directory /
 
+# p4-fusion installation
+# Fetching p4 sources archive
+RUN wget https://github.com/salesforce/p4-fusion/archive/refs/tags/v1.2.zip && \
+    mv v1.2.zip /usr/local/bin && \
+    unzip /usr/local/bin/v1.2.zip -d /usr/local/bin
```
Comment (ryanslade) [inline cmd/gitserver/Dockerfile:63]: Maybe we can unzip into a dir without the version number? Ideally when we need to update this to support a new version we should only need to change the URL of the archive we need to download.

File: cmd/gitserver/Dockerfile (hunk 2 above)
Comment (sashaostrikov) [reply cmd/gitserver/Dockerfile:87]: yes, I am already doing it in the second part of the command :)
`&& rm -rf /usr/local/bin/p4-fusion-1.2`

File: cmd/gitserver/Dockerfile (hunk 1 above)
Comment (sashaostrikov) [reply cmd/gitserver/Dockerfile:61]: thanks! I'll pick 1.3

File: cmd/gitserver/Dockerfile (hunk 4 above)
Comment (sashaostrikov) [reply cmd/gitserver/Dockerfile]: nice point, fixed it in my recent commit

File: cmd/gitserver/Dockerfile (hunk 1 above)
Comment (sashaostrikov) [reply cmd/gitserver/Dockerfile]: I even did 1.4 (and 1.5 is coming today as well, so I'll switch to it later)

File: cmd/gitserver/Dockerfile [hunk 5]
```diff
@@ -55,10 +57,43 @@ COPY --from=coursier /usr/local/bin/coursier /usr/local/bin/coursier
...
+# p4-fusion installation completed here
+
 RUN mkdir -p /data/repos && chown -R sourcegraph:sourcegraph /data/repos
 USER sourcegraph
 
 WORKDIR /
 
-ENTRYPOINT ["/sbin/tini", "--", "/usr/local/bin/gitserver"]
-COPY gitserver /usr/local/bin/
+#ENTRYPOINT ["/sbin/tini", "--", "/usr/local/bin/gitserver"]
```
Comment (ryanslade) [inline cmd/gitserver/Dockerfile]: Don't forget to uncomment this 😄

**Comment by sashaostrikov** [comment]
PTAL

now it is v1.5 and everything launches successfully

File: cmd/gitserver/Dockerfile (hunk 5 above)
Comment (sashaostrikov) [reply cmd/gitserver/Dockerfile]: yep, just did it :)

**Review by ryanslade — APPROVED** [review]
LGTM!

**Pull request merged**